### PR TITLE
mds: support constness in MDSCacheObjects

### PR DESCRIPTION
**Pull request by jcsp**
So that one can have const CInode and CDir references
from time to time.

This is mostly just dumb adding 'const' to declarations
and iterators, but there is also an awkward case in CInode
where CINode::get_caps_issued was resetting loner_cap to -1
based on get_auth(), so that has changed functionally to
update loner_cap in state_set/state_clear/state_reset
instead so that the getter is just a getter.

Signed-off-by: John Spray <john.spray@redhat.com>

### CODE REVIEW

File: src/mds/CInode.cc [hunk 1]
```diff
@@ -2808,13 +2808,13 @@ int CInode::get_caps_allowed_for_client(client_t client)
...
 int CInode::get_caps_issued(int *ploner, int *pother, int *pxlocker,
-			    int shift, int mask)
+			    int shift, int mask) const
 {
   int c = 0;
   int loner = 0, other = 0, xlocker = 0;
-  if (!is_auth())
-    loner_cap = -1;
-  for (map<client_t,Capability*>::iterator it = client_caps.begin();
+  assert(is_auth() || loner_cap == -1);
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: This looks good to me, but how'd you check it?

File: src/mds/CInode.h [hunk 2]
```diff
@@ -748,18 +790,21 @@ class CInode : public MDSCacheObject, public InodeStore {
...
       return client_caps[client];
     return 0;
   }
-  int get_client_cap_pending(client_t client) {
-    Capability *c = get_client_cap(client);
-    if (c) return c->pending();
-    return 0;
+  int get_client_cap_pending(client_t client) const {
+    if (client_caps.count(client)) {
+      return client_caps.find(client)->second->pending();
```
Comment (gregsfortytwo) [inline src/mds/CInode.h]: It's a silly little thing, but there are some static analyzers that will complain about our not using iterators here — makes life harder on the compiler when optimizing.

File: src/mds/Server.cc [hunk 3]
```diff
@@ -7211,6 +7211,7 @@ void Server::do_rename_rollback(bufferlist &rbl, mds_rank_t master, MDRequestRef
   inode_t *pi = 0;
   if (in) {
     if (in->authority().first == whoami) {
+      inode_t *
```
Comment (gregsfortytwo) [inline src/mds/Server.cc]: Ummm, what? No shadowing, please!

File: src/mds/CInode.h [hunk 4]
```diff
@@ -702,17 +725,36 @@ class CInode : public MDSCacheObject, public InodeStore {
...
+    if (!is_auth()) {
+      loner_cap = -1;
+    }
+  }
+  void state_reset(unsigned s) {
+    MDSCacheObject::state_reset(s);
+    if (!is_auth()) {
+      loner_cap = -1;
+    }
+  }
```
Comment (gregsfortytwo) [inline src/mds/CInode.h]: Help me out here; why do we need to do this? Could it be broken out into a separate patch? :)

File: src/mds/CInode.h (hunk 2 above)
Comment (jcsp) [inline src/mds/CInode.h]: Done

File: src/mds/Server.cc (hunk 3 above)
Comment (jcsp) [inline src/mds/Server.cc]: Ah, that was just a random typo, good catch!

**Comment by jcsp** [comment]
There were a few places setting `state` directly, so rather than introduce the possibility of subtle bugs I've just removed the set_state stuff on CInode, and made get_caps_issued non-const.  We can always come back to refactoring that in future.
